### ISSUE TITLE
[claude-review] Correctness bugs in numba processor kernels (severity: critical)

### ISSUE THREAD

**Issue by gipert** [legend-exp/dspeed#205]
## How this review was generated

This issue is part of an automated code review of `dspeed` performed with Claude Code (Claude Fable 5) on 2026-07-02, at the request of a maintainer. Five parallel read-only review agents examined separate areas of the codebase (processing chain core, `build_dsp`/CLI/visualization, processors A-L, processors M-Z, packaging/CI/tests). The coordinating session then re-verified every critical and high severity claim directly against the source, compiling and running kernels in a local venv where relevant; candidate findings that failed verification were dropped. Issues are grouped by topic so that each can be assigned to a dedicated agent or contributor, and are sorted by severity.

**Overall severity: critical**

This group covers correctness defects in the numba processor kernels under `src/dspeed/processors/`. Line numbers refer to current `main` (a876536).

---

### 1. `get_multi_local_extrema`: conservative-AND search fills minima output from the maxima arrays (critical)

`src/dspeed/processors/get_multi_local_extrema.py:255-256`

In the `search_direction == 2` branch, the minima block was copied from the maxima block without swapping the variables:

```python
# Do the same for the minima
rge_right = (right_vt_max[~np.isnan(right_vt_min)]).astype(np.int_)
rge_left = (left_vt_max[~np.isnan(left_vt_min)]).astype(np.int_)
```

It masks `right_vt_max`/`left_vt_max` (maxima indices) using the NaN pattern of the minima arrays, and the result is written into `vt_min_out`/`n_min_out`. Whenever the directional searches find different numbers of maxima and minima (the normal case for asymmetric waveforms), `vt_min_out` contains maxima indices or a wrong count. The existing tests for this branch (`test_get_multi_local_extrema_both_cons*`) only assert on the maxima outputs, so the bug is not caught.

**Fix:** use `right_vt_min`/`left_vt_min` in those two lines, matching the pattern of the maxima block above and of the `search_direction == 3` branch. Extend the tests to assert on `vt_min_out`/`n_min_out`.

### 2. `saturation`: wrong high-saturation threshold formula (critical)

`src/dspeed/processors/saturation.py:68`

```python
elif w_in[i] == np.power(2, int(bit_depth_in)) - int(bit_depth_in):
```

For an N-bit ADC the saturated code is `2**N - 1`, not `2**N - N`. For `bit_depth_in=16` this compares against 65520 instead of 65535, so genuinely clipped samples are essentially never counted and `n_hi_out` is silently wrong. There is no `test_saturation.py`.

**Fix:** compare against `np.power(2, int(bit_depth_in)) - 1` and add a unit test.

### 3. `histogram` / `histogram_around_mode`: the maximum sample is silently dropped (high)

`src/dspeed/processors/histogram.py:84-88`, `:180-184`, `:200-208`

The binning loop uses a strict `<` against the upper bin edges, and the last border equals `max(w_in)`. A sample equal to the maximum (always present, since the edges are derived from the data) satisfies no bin condition and is discarded. Standard histogram semantics (`numpy.histogram`) close the last bin on the right. The existing test `test_histogram_fixed_width` encodes the bug: it feeds 100 samples and asserts a bin pattern summing to 99.

**Fix:** clamp the bin index so the maximum lands in the last bin (for example compute `k = min(int((w_in[i]-wf_min)/delta), n_bins-1)` directly), and fix the test expectation.

### 4. `inject_damped_oscillation`: output left uninitialized on NaN input (high)

`src/dspeed/processors/inject_ringing.py:47-48`

Unlike every other kernel in the package, this function does not set `w_out[:] = np.nan` before its early `return` on NaN input. Since `w_out` is a pre-allocated, reused buffer, an event with a NaN sample yields stale garbage from the previous event in the output instead of NaN, producing plausible-looking but wrong waveforms.

**Fix:** add `w_out[:] = np.nan` as the first statement.

### 5. `bi_level_zero_crossing_time_points`: crossing at index 0 is silently dropped (high)

`src/dspeed/processors/time_point_thresh.py:490-533`

`is_above_thresh`/`is_below_thresh` start as `False` and are later assigned a waveform index (`is_above_thresh = i`), then used both as truthiness flags (`if is_above_thresh and ...`) and in arithmetic (`i - is_above_thresh`). If a crossing is recorded at index 0 (possible when `t_start_in == 0`), the stored value is falsy and later checks treat it as "not yet crossed", dropping a legitimate zero crossing at the start of the search window.

**Fix:** use separate boolean flags alongside the stored indices instead of overloading one variable for both roles.

### 6. `peak_snr_threshold`: search window excludes the right edge (high)

`src/dspeed/processors/peak_snr_threshold.py:58-66`

With `a = idx - width`, `b = idx + width`, the loop `for j in range(a, b, 1)` scans `[a, b-1]`, so the sample at `idx + width` is never examined as a minimum candidate (note `b` is clamped to `len(w_in) - 1`, which is a valid index). If the true local minimum sits exactly at the right edge, the amplitude-ratio test uses a wrong minimum and can reject a genuine peak.

**Fix:** iterate `range(a, b + 1, 1)` to make the window symmetric and inclusive.

### 7. `get_wf_centroid`: unguarded `IndexError` instead of NaN/DSPFatal (medium)

`src/dspeed/processors/get_wf_centroid.py:62-67`

```python
c_a = np.where(w_in[w_in.argmin() : w_in.argmax()] > 0)[0][0] + ...
c_b = np.where(w_in[w_in.argmin() : w_in.argmax()] < 0)[0][-1] + ...
```

This assumes `argmin() < argmax()` and that a zero crossing exists in between. For reversed polarity (empty slice) or when no sample satisfies the condition, `np.where(...)[0]` is empty and the `[0]`/`[-1]` indexing raises a raw `IndexError` out of the kernel, instead of the NaN/`DSPFatal` convention used elsewhere. Note this is extra dangerous in combination with the `ProcessingChain.execute()` bug reported in the companion processing-chain issue, where a stray `IndexError` is interpreted as end-of-data.

**Fix:** check for empty results (and slice orientation) and set the outputs to NaN or raise `DSPFatal` with a clear message.

### 8. `avg_current`: `length == 0` passes validation but crashes on the negative-zero slice (medium)

`src/dspeed/processors/moving_windows.py:243-249`

The validation explicitly allows `length == 0`, but `w_in[: -int(length)]` evaluates as `w_in[:0]` (empty) because `-0 == 0`, producing a broadcasting error; the subsequent `w_out /= length` would also divide by zero.

**Fix:** require `length >= 1` in the validation (or special-case 0), or rewrite the slice as `w_in[: len(w_in) - int(length)]`.

### 9. `ml.softmax` is actually softplus, and its stability claim is false (medium)

`src/dspeed/processors/ml.py:73-80`

The function named `softmax` computes `np.log(1 + np.exp(x_in))`, which is softplus, not softmax, and the docstring claims it is "numerically stable" when it overflows to `inf` for inputs above ~88 (float32) / ~709 (float64).

**Fix:** rename to `softplus` (updating the `'m'` activation references in the `dense_layer_*`/`classification_layer_*` docstrings) and use the stable form `np.maximum(x_in, 0) + np.log1p(np.exp(-np.abs(x_in)))`.

### 10. `tf_model`: only `OSError` is caught when loading the Keras model (low)

`src/dspeed/processors/tf_model.py:33-51`

`tf.keras.models.load_model` can also raise `ValueError`/deserialization errors for malformed or incompatible model files; those bypass the `tfmodel_null` fallback the `except OSError` branch is meant to provide.

**Fix:** broaden the except clause or document that only missing files are handled gracefully.

### 11. `cusp_filter`/`zac_filter`: flat top is one sample longer than `flat` (low, needs author confirmation)

`src/dspeed/processors/energy_kernels.py:67-70` and `:137-140`

`for ind in range(lt, lt + flat_int + 1, 1): kernel[ind] = 1` sets `flat + 1` samples, one more than the documented flat-section length. The pattern is identical in both filters, so it may be an intentional symmetry convention, but it contradicts the docstring and no test pins the exact flat length.

**Fix:** if unintentional, drop the `+ 1`; if intentional, document that the flat section has `flat + 1` samples.

---

Note: three additional candidate findings (claimed numba `TypingError`s from float arguments to `range()` in `presum`, `inject_gumbel` and `soft_pileup_corr`) were disproved by compiling and running the kernels, and are not included. However, none of `saturation`, `presum`, `pmt_pulse_injector`, `soft_pileup_corr` or `peak_snr_threshold` has a unit test; adding coverage there would be a good follow-up.

**Comment by iguinn** [comment]
@ggmarshall @willquinn @sofia-calgaro can you look into issue 2 about the pileup cut? My intuition agrees with claude here, that it should be `2^N - 1`; it also seems like a `>=` check is more appropriate here than `==`. Do we know if this processor works correctly for us?

@schwarzmario can you look into issue 3 about the histogram dropping the max? Note that in the performance review, it also suggested some changes that I implemented, but I left it dropping the max. Is this behavior preferred, or do we want to add the max to the maximum bin?

@valerioda can you look into issue 11? Is this flat top length intended?

Issue 4 is fixed by https://github.com/legend-exp/dspeed/pull/213
